index.rst

```diff
@@ -15,15 +15,123 @@ Abstract
+Cassandra supports secondary indexing, but its performance is not adequate for replication use case.
+One of the ideas suggested at `PPDB architecture meeting`_ was to write new DIA records to a separate set of files and transfer those files asynchronously to PPDB.
+This approach would cause many complications, so the alternative solution was designed that uses the same Cassandra storage.
+
+The new records, as they are inserted into regular DIA tables, are also stored in separate set of staging tables that use time-based partitioning and indexing.
+Every store operation that inserts a number of records into APDB is identified by an ``insert ID`` token.
+The replication process will discover new ``insert ID`` tokens in APDB, copy the data corresponding to that token to PPDB, and remove that token and its corresponding records from APDB.
+APDB has special table that tracks newly-created ``insert ID`` tokens, PPDB (or a separate replication database) will probably need similar approach to track tokens that have been replicated.
+
+The ordering of transfers in replication process is important.
+For optimal performance the DIAObject records in Cassandra implementation do not have their ``validityEnd`` column updated, so it always is always ``NULL``.
+When DIAObjects are copied to PPDB, the replication process has to populate ``validityEnd`` of the DIAObject which is already in PPDB using ``validityStart`` of the same DIAObject being replicated.
+This implies that the time order of the replication needs to be the same as time order in which data were originally inserted into APDB, without such ordering it will be very difficult to manage ``validityEnd`` updates.
+
+To implement propagation of PPDB updates to APDB the replication process will need to know which DIASources need to be re-associated and which DIAObjects need to be removed.
+To support this, PPDB will probably use similar approach of keeping the separate log of updates that can be replayed on APDB side.
+
+At this point APDB already implements all operations needed to support replication process, but it was never tested as we do not have PPDB instance or its corresponding API yet.
+
+
+Replicating non-DIA tables
+==========================
+
+New SSSource and SSObject records produced by AP pipeline need to be copied to PPDB along with DIA tables.
+It would be reasonable to use the same mechanism relying on ``insert ID`` tokens and separate staging tables for the new records.
+Special care will be needed in the replication process to order PPDB updates, as SSSource records have a dependency on their corresponding DIASource records and some DIASources are linked to SSObjects (it may not be expressed explicitly in the current ``sdm_schema`` definition, but it should be).
+Additionally, as some DIASources can be embargoed for a period of time or permanently, the corresponding SSSource and SSObject records replication has to be delayed as well.
```

Comment:
If we are building SSSource and SSObject from the PPDB, and the PPDB contains only non-embargoed data, then no embargoed DIASources can be part of an SSSource or SSObject.  So there should be no need for delayed replication.

Reply:
I was thinking about possible delays in embargo decisions. Do we always get an immediate decision or could happen that we are told - "hold it for now and ask again tomorrow"? Or something like "do not "use it until three days later"?